File: 3.longest-substring-without-repeating-characters.go
/*
 * @lc app=leetcode id=3 lang=golang
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
package main

func lengthOfLongestSubstring(s string) int {
	if len(s) <= 1 {
		return len(s)
	}
	var ele2Idx map[rune]int = map[rune]int{rune(s[0]): 0}
	maxLen := 1
	left := 0
	for i, c := range s {
		if i == 0 {
			continue
		}
		neoLeft, has := ele2Idx[c]
		if has {
			neoLen := i - left
			if neoLen > maxLen {
				maxLen = neoLen
			}
			for ; left <= neoLeft; left++ {
				delete(ele2Idx, rune(s[left]))
			}
		}
		ele2Idx[c] = i
	}
	if len(ele2Idx) > maxLen {
		return len(ele2Idx)
	} else {
		return maxLen
	}
}

// @lc code=end

File: 3.longest-substring-without-repeating-characters.java
import java.util.HashMap;
import java.util.Map;

/*
 * @lc app=leetcode id=3 lang=java
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int res = 1;
        if (s == null || s.length() == 0) {
            return 0;
        }
        Map<Character, Integer> memo = new HashMap<>(52);
        int lastStart = 0;
        for (int i = 0; i < s.length(); ++i) {
            char c = s.charAt(i);
            if (memo.containsKey(c)) {
                int lastLen = i - lastStart;
                res = Math.max(res, lastLen);
                int newStart = memo.get(c) + 1;
                for (; lastStart < newStart; ++lastStart) {
                    memo.remove(s.charAt(lastStart));
                }
            }
            memo.put(c, i);
        }
        return Math.max(res, memo.size());
    }
}
// @lc code=end
